### ISSUE TITLE
`weave` and `weave --local` behave differently

### ISSUE THREAD

**Issue by awh** [weaveworks/weave#507]
`weave` behaves differently depending on whether `--local` is specified:

### `STDERR` is merged into `STDOUT` when run remotely
````
vagrant@ubuntu-14:~/weave$ sudo ./weave status 2>/dev/null
weave container is not present; have you launched it?

weavedns container is not present; have you launched it?
vagrant@ubuntu-14:~/weave$ sudo ./weave --local status 2>/dev/null

vagrant@ubuntu-14:~/weave$
````
### Lines are terminated with CRLF when run remotely
````
vagrant@ubuntu-14:~/weave$ sudo ./weave ps|hexdump -c
0000000   b   b   7   e   6   d   a   0   b   e   1   4       3   6   :
0000010   1   1   :   0   4   :   9   8   :   e   e   :   a   6       1
0000020   0   .   0   .   0   .   1   /   2   4  \r  \n   a   d   1   9
0000030   1   8   0   7   7   8   0   d       7   a   :   b   c   :   a
0000040   c   :   d   3   :   0   4   :   f   f  \r  \n
000004c
vagrant@ubuntu-14:~/weave$ sudo ./weave --local ps|hexdump -c
0000000   b   b   7   e   6   d   a   0   b   e   1   4       3   6   :
0000010   1   1   :   0   4   :   9   8   :   e   e   :   a   6       1
0000020   0   .   0   .   0   .   1   /   2   4  \n   a   d   1   9   1
0000030   8   0   7   7   8   0   d       7   a   :   b   c   :   a   c
0000040   :   d   3   :   0   4   :   f   f  \n
000004a
vagrant@ubuntu-14:~/weave$
````

**Comment by awh** [comment]
The LF->CRLF termination is being caused by the intervening Docker pseudoterminal. Can you see any problems with changing the `weaveexec/Dockerfile` entrypoint to the following @squaremo?

    ENTRYPOINT ["stty -onlcr; /home/weave/weave"]

**Comment by squaremo** [comment]
I think it would have to be

```
ENTRYPOINT "stty -onlcr; /home/weave/weave"
```

but otherwise, worth a crack yep.

**Comment by dpw** [comment]
I suspect neither of those will work as presented due to the need to handle arguments passed to `docker run`.  The docker docs are not very clear on the matter though.

**Comment by awh** [comment]
You're right, it requires the shell form of entrypoint, and consequently can't be used for two reasons:

1. Shell form doesn't permit arguments to be passed through from `docker run`
2. Shell form doesn't allow signals to pass through

**Comment by awh** [comment]
How about we add an intervening wrapper script inside the exec container e.g:
````
#!/bin/sh
stty -onlcr
exec /home/weave/weave
````
and have that as the entrypoint?

**Comment by dpw** [comment]
I don't think there's any huge obstacle to crafting an appropriate entrypoint.  It's just not quite as simple as the ones shown.  I'm not sure this is right, but it's probably closer:

    ENTRYPOINT ["sh", "-c", "stty -onlcr; exec /home/weave/weave \"$@\"", "dummy"]

**Comment by awh** [comment]
Yes that has exactly the desired effect.

On the stdout/stderr merging issue - it looks like Docker is not able to keep these separate when the `-t` option is specified: docker/docker#8040. Consequently removing `-t` from `exec_remote` fixes both issues, but then signal handling will break. Thoughts @bboreham @rade?

**Comment by awh** [comment]
This flag to `docker run` looks promising as a way of preserving signal delivery without using `-t`:
````
--sig-proxy=true: Proxify all received signal to the process (non-TTY mode only)
````
Unfortunately it doesn't seem to work!

**Comment by awh** [comment]
Further investigation reveals the reason why it doesn't work. The signals *are* being delivered, but as any PID 1 process (even in a namespace) has `SIGNAL_UNKILLABLE` the default signal handler does not terminate the process. It is however possible for the contained process to install its own signal handler and exit voluntarily, unfortunately most shells mask signals whilst waiting for subprocesses to exit so we can't simply use `trap` before commencing the weave script. Instead a small custom executable is required that executes the weave script and propagates `SIGINT` and other desired signals - once this is done we can remove the `-t` option thereby fixing the issues outlined above.

**Comment by bboreham** [comment]
Replacing `-t` with `--sig-proxy=false` is a possible stop-gap.  It still doesn't kill the underlying process, but avoids the munging of output that comes with the pty.

**Comment by bboreham** [comment]
My reading of the kernel source suggests that SIGNAL_UNKILLABLE relates specifically to SIGKILL, rather than SIGINT.  Suggest that further research is required if we want to know the true reason.

**Comment by bboreham** [comment]
I changed my mind: [the source](https://github.com/torvalds/linux/blob/4f671fe2f9523a1ea206f63fe60a7c7b3a56d5c7/kernel/signal.c#L74) says that Linux ignores signals with default action when SIGNAL_UNKILLABLE is set.

**Comment by rade** [comment]
Does the `-t` we have on master actually work?

I've added a sleep with
````patch
diff --git a/weave b/weave
index 0eb30ed..1abd2ac 100755
--- a/weave
+++ b/weave
@@ -76,6 +76,7 @@ if [ "$1" = "--local" ]; then
         echo "weave must be run as 'root' when run locally" >&2
         exit 1
     }
+    sleep 10
 elif [ "$1" != "run" ]; then
     exec_remote "$@"
     exit $?
````
rebuild, run `weave launch` and ctrl-c that. This does not stop the weavexec container and the weave container still gets launched. By contrast, this does work as expected with the changes from #523.

So does this issue in fact cover *three* issues with remote execution?

1. stderr is merged into stdout
2. line termination with CRLF instead of LF
3. ctrl-c is ignored

**Comment by bboreham** [comment]
My experiments, prompted by @awh, showed that ctrl-c with `-t` option caused the Docker client to be killed but not the running container.
